plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/conversions/RemoveUnnecessaryParenthesesConversion.kt

```diff
@@ -0,0 +1,213 @@
+        if (expression.children.isEmpty()) return false
+        if (expression.children.filterIsInstanceAnd<JKExpression> { it.containsNewLine() }.isNotEmpty()) return true
+        return expression.children.filterIsInstance<JKExpression>().any { recursivelyContainsNewlines(it) }
+    }
+
+    private fun recursivelyContainsNewlineBeforeOperator(expression: JKBinaryExpression): Boolean {
+        val left = expression.left
+        val right = expression.right
+        if (recursivelyEndsWithNewline(left)) {
+            return true
+        }
+        if (left is JKBinaryExpression && recursivelyContainsNewlineBeforeOperator(left)) return true
+        if (right is JKBinaryExpression && recursivelyContainsNewlineBeforeOperator(right)) return true
+        return false
+    }
+
+    private fun recursivelyEndsWithNewline(expression: JKExpression): Boolean {
+        if (expression.hasLineBreakAfter) return true
+        val lastChild = expression.children.lastOrNull()?.safeAs<JKExpression>() ?: return false
+        return recursivelyEndsWithNewline(lastChild)
+    }
+
+    private fun recursivelyStartsWithNewline(expression: JKExpression): Boolean {
+        if (expression.hasLineBreakBefore) return true
+        val lastChild = expression.children.firstOrNull()?.safeAs<JKExpression>() ?: return false
+        return recursivelyStartsWithNewline(lastChild)
+    }
+
+    /**
+     * Based on `getPriority` in KtPsiUtil.java
+     */
+    private fun getPriority(expression: JKElement): Int {
+        val maxPriority = KotlinExpressionParsing.Precedence.entries.size + 1
+
+        if (expression is JKPostfixExpression || expression is JKQualifiedExpression || expression is JKCallExpression || expression is JKMethodReferenceExpression) {
+            return maxPriority - 1;
+        }
+
+        if (expression is JKPrefixExpression || expression is JKLabeledExpression) return maxPriority - 2
+
+        if (expression is JKIfElseExpression) return KotlinExpressionParsing.Precedence.ASSIGNMENT.ordinal
+
+        if (expression is JKSuperExpression) return maxPriority
+
+        if (expression !is JKBinaryExpression) return maxPriority
+
+        val operatorToken = expression.operator.token
+        val operatorElementType = JKOperatorToken.toKtElementType(operatorToken)
+            ?: return if (operatorToken.text in STD_INFIX_FUNCTIONS) KotlinExpressionParsing.Precedence.SIMPLE_NAME.ordinal else maxPriority
+
+        for (precedence in KotlinExpressionParsing.Precedence.entries) {
+            if (precedence != KotlinExpressionParsing.Precedence.PREFIX &&
+                precedence != KotlinExpressionParsing.Precedence.POSTFIX &&
+                (precedence.operations.contains(operatorElementType))
+            ) {
+                return maxPriority - precedence.ordinal - 1
+            }
+        }
+
+        return maxPriority
+    }
+
+    companion object {
+        val STD_INFIX_FUNCTIONS = setOf(
```

Comment:
Do we need all of them? Is it enough to check that the operator is `JKKtWordOperatorToken`?